### DISCUSSION TITLE
Force Images to be cropped to a certain aspect ratio

### DISCUSSION THREAD

**Discussion by MiniDigger** [directus/directus#3286] [Feature Requests]
Two (well, 3) different requests:

* configure the possible aspect ratios
* configure the allowed aspect ratios for a field in a collection 
* force the user to crop to those allowed ratios on upload

the usecase for this would be a self service to edit certain images on a website (for example, profile pic), where images have to be of a certain aspect ratio, for them to be displayed properly.

else they would need to be cropped automatically by software and we would loose the point of interest in the image.

**Comment by rijkvanzanten** [comment]
To confirm, this would be during upload of the file right? Not when accessing it through `/assets`?

**Reply by MiniDigger** [reply]
yes, on upload. basically a restriction on the stored image.

**Reply by cynova** [reply]
We need this aswell.
We started implementing it in Directus V8, but will probably abandon it now that Directus 9 is advancing so quickly.

The challenging part seems to be how to handle edits. If someone were to attach the Image to a collection item forcing an aspect ratio, then we get a conflict if someone tries to crop the image (from files page or from another collection).
This is generally a problem even without enforcing aspect ratios (side effects when editing Files, be it the files themselves or Title etc.)
A possible solution would be to add a "Immutable Relation" restriction that causes all Edits to a file attached to a collection with said restriction to fail. This would also automatically prevent cache related issues when changing image aspect ratio.

What are your thoughts on this @rijkvanzanten ?

**Comment by TheAzack9** [comment]
I was writing a separate idea on this, but it seems i was just able to find this before posting it :) A huge +1 from me! Would be a very nice feature to have.

It would also be nice if we were able to specify an min width. min height, max width, max height restrictions aswell. This way you can force the image to be a specific size if necessary aswell.

**Comment by albrrt** [comment]
I also agree - this feature would be greatly appreciated!

Same way you can restrict a text field to a certain character count - you should be able to restrict the image field to a certain width/height and if the image you upload does not match these requirements you can crop it "on-the-go"

**Comment by paulboudewijn** [comment]
I'd also like to have all the features mentioned, but maybe the following is relatively easy to implement and makes the image editor a lot more usable right away:

1. customizable aspect ratio's (non field specific)
2. extend image interface with properties: 'allow dragndrop', 'allow browse', 'allow import', 'allow library'
3. extend image interface with property 'allow access to the parent folder'

Since image editing is only available when uploading through the file library, configuring an image field to only allow choosing a library image will force a user to choose a pre-cropped image. With folder names similar to custom aspect ratio names, it should be pretty easy to teach users where to save their cropped images (i.e. put all your hero images in the '/hero-images' folder and crop them using the 'hero image' aspect ratio).
Then we could choose the hero-images folder as starting folder for a 'hero-image' field and we could disable access to the parent folder so that users can't browse to a folder intended to store images with a different aspect ratio. 

It's not 100% foolproof, of course, but I think it will definitely work for my clients.

**Comment by jawngee** [comment]
I have a fork of directus 9 we are using for a client site has the configurable aspect ratio presets functionality and it also adds a toggle so that any crop you perform creates a new image instead of cropping the existing one.  I can extract that functionality to a new fork that I can put on github if someone wants to run with it.

**Reply by paulboudewijn** [reply]
That would be great!

**Reply by jawngee** [reply]
I will get it up by next Tuesday at the latest.

**Comment by jawngee** [comment]
Here's my commit to a fork that adds crop presets:  https://github.com/jawngee/directus/commit/650600ca0002f21797135307170dcfff4aadbcb8

You'll need to add a json file to the root of your api app.  It's included in the commit, you can modify this add or remove presets.  The one problem is that you can't supply a custom icon for a crop preset.

Use it or use it as a reference, hope it helps.

**Reply by paulboudewijn** [reply]
Many thanks!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
